vector/v.fill.holes/main.c

```diff
@@ -0,0 +1,224 @@
+            !Vect_cats_in_constraint(area_cats, field, constraint_cat_list)) {
+            continue;
+        }
+        Vect_write_line(&output, GV_CENTROID, points, area_cats);
+
+        Vect_get_area_points(&input, area, points);
+        Vect_write_line(&output, GV_BOUNDARY, points, empty_cats);
+
+        if (field > 0) {
+            Vect_field_cat_get(area_cats, field, field_cats);
+            Vect_list_append_list(all_cats, field_cats);
+        }
+    }
+
+    Vect_destroy_cats_struct(empty_cats);
+    Vect_destroy_cats_struct(area_cats);
+    Vect_destroy_line_struct(points);
+
+    /* Let's get vector layers db connections information */
+    struct field_info *input_info = NULL;
+
+    if (field > 0 && all_cats->n_values)
+        input_info = Vect_get_field2(&input, options.field->answer);
+
+    if (input_info) {
+        G_verbose_message(_("Copying attributes for layer <%s>"),
+                          options.field->answer);
+
+
+        /* outputput information useful for debuging 
+           include/vect/dig_structs.h
+         */
+        G_debug(1,
+                "Field number:%d; Name:<%s>; Driver:<%s>; Database:<%s>; Table:<%s>; Key:<%s>;\n",
+                input_info->number, input_info->name, input_info->driver,
+                input_info->database, input_info->table, input_info->key);
+
+        struct field_info *output_info =
+            Vect_default_field_info(&output, field, NULL, GV_1TABLE);
+        /* Create database for new vector map */
+        dbDriver *driver = db_start_driver_open_database(output_info->driver,
+                                                         output_info->
+                                                         database);
+
+        G_debug(1,
+                "Field number:%d; Name:<%s>; Driver:<%s>; Database:<%s>; Table:<%s>; Key:<%s>;\n",
+                output_info->number, output_info->name, output_info->driver,
+                output_info->database, output_info->table, output_info->key);
+
+        /*
+           Vect_map_add_dblink(&output, output_info->number, output_info->name,
+           output_info->table, input_info->key,
+           output_info->database, output_info->driver);
+         */
+
+        /* Copy attribute table data */
+        if (db_copy_table_by_ints
+            (input_info->driver, input_info->database, input_info->table,
+             output_info->driver, Vect_subst_var(output_info->database,
+                                                 &output), output_info->table,
+             input_info->key, all_cats->value,
+             all_cats->n_values) == DB_FAILED)
+            G_fatal_error(_("Unable to copy attribute table to vector map <%s>"),
+                          options.output->answer);
+        db_close_database_shutdown_driver(driver);
+    }
+
+    Vect_destroy_list(field_cats);
+    Vect_destroy_list(all_cats);
+
+    Vect_build(&output);
+
+    Vect_set_release_support(&input);
+    Vect_set_release_support(&output);
```

Comment:
Not needed, the operating system does it better and faster.

Reply:
I tested the performance on a vector with 186,940 areas (180,015 after the operation). Here are its stats:

```
nodes=227263
boundaries=352424
centroids=166520
areas=186940
islands=61779
primitives=518944
```

I used perf stat for time and valgrind for the leaked memory. Here are the results:

| memory release | perf stat -r 10 seconds time elapsed (3x10 runs) | valgrind LEAK SUMMARY still reachable; definitively lost |
| --- | --- | --- |
| enabled | 24.39, stddev 0.18 (24.181, 24.488, 24.513) | 8,297,834 bytes in 145 blocks; 191,032 bytes in 184,529 blocks |
| disabled | 23.97, stddev 0.36 (23.557, 24.212, 24.151) | 248,489,855 bytes in 6,580,166 blocks blocks; 6,579 bytes in 76 blocks |
| difference (enabled minus disabled) | 0.42 (1.7%) | -240,192,021 (regression) ; 177,950 (improvement)

I don't think the results here support the idea that OS will be "much faster" as the old [doc/debugging.txt](https://github.com/OSGeo/grass/blob/main/doc/debugging.txt) claims in a note from 2005 (https://github.com/OSGeo/grass-legacy/commit/fb92c0fd2163474f18daf0413d62e89e28249fec). The difference of averages is smaller than 2*stddev.

However, given the overall lack of memory handling and a policy to tackle that, I'm fine leaving that out for the small performance gain (for now).